feat: ROOT-45: Filter by annotator together with annotation results

matt-bernstein opened the pull request:
<!-- Follow Merge -->
This PR was automatically generated via Follow Merge. [Workflow Run](https://github.com/HumanSignal/label-studio/actions/runs/16815444605)
Please ensure that all linked upstream Pull Requests are merged before proceeding with this one.
- https://github.com/HumanSignal/label-studio-sdk/pull/499
<!-- Follow Merge -->

<img width="593" height="266" alt="image" src="https://github.com/user-attachments/assets/c5406b0d-ba96-4382-9787-1ad67397756f" />


Allow filters in DM to have a `child_filter` which is only available after the parent is selected. Use this new capability to add `annotator` as a `child_filter` to `annotation_results`. In this example, filters are now expressive enough to search for tasks with an annotation with label "PER" created by annotator "matt". Previously, a false positive would be returned if a task had both an annotation with label "PER" *and* a (potentially different) annotation by annotator "matt".

**FE changes**
- give filters a `child_filter` and handle conditional CRUD logic (ensure a child never exists without its parent, and that incomplete filters are never sent to the BE)
- styling to join parent and child filters into a visually distinct "filter line"
- 
**BE changes**
- let the Filter model have a recursive "children" field
- expose children in serializers, limiting to at most 1 child and 1 level of nesting
- ensure "filter lines" are always AND-ed when applying them to the tasks queryset, no matter what the FilterGroup's outer conjunction is
- 
**LSE changes**
- the `annotator` child_filter
- tests



<!--

This description MUST be filled out for a PR to receive a review. Its primary purposes are:

 - to enable your reviewer to review your code easily, and
 - to convince your reviewer that your code works as intended.

Some pointers to think about when filling out your PR description:
 - Reason for change: Description of problem and solution
 - Screenshots: All visible changes should include screenshots.
 - Rollout strategy: How will this code be rolled out? Feature flags / env var / other
 - Testing: Description of how this is being verified
 - Risks: Are there any known risks associated with this change, eg to security or performance?
 - Reviewer notes: Any info to help reviewers approve the PR
 - General notes: Any info to help onlookers understand the code, or callouts to significant portions.

You may use AI tools such as Copilot Actions to assist with writing your PR description (see https://docs.github.com/en/copilot/using-github-copilot/using-github-copilot-for-pull-requests/creating-a-pull-request-summary-with-github-copilot); however, an AI summary isn't enough by itself. You'll need to provide your reviewer with strong evidence that your code works as intended, which requires actually running the code and showing that it works.

-->

matt-bernstein committed:
feat: ROOT-45: Filter by annotator prototype

matt-bernstein committed:
model changes

matt-bernstein committed:
migration

matt-bernstein committed:
update migration

matt-bernstein committed:
Revert "update migration"

This reverts commit 21be28fee861bbd95ca593ff3750e79e85fe3cff.

matt-bernstein committed:
async migration

matt-bernstein committed:
give up on index

matt-bernstein committed:
omit parent when appropriate

matt-bernstein committed:
wip FE for child filters

matt-bernstein committed:
allow parent index instead of parent id

matt-bernstein committed:
use parent_index

matt-bernstein committed:
ff

matt-bernstein committed:
bugfix

matt-bernstein committed:
filter counter bugfix

matt-bernstein committed:
don't double-display child filters

matt-bernstein committed:
apply child filters correctly

matt-bernstein commented:
/fmt
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15740108174)

matt-bernstein committed:
Apply pre-commit linters

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15740108174

matt-bernstein committed:
extend pydantic data model

jombooth commented on label_studio/data_manager/models.py:
```diff
@@ -86,7 +86,25 @@ class FilterGroup(models.Model):
+        related_name='children',
+        null=True,
+        blank=True,
+        db_index=False,
```
Is this to satisfy the linter? While it's strictly necessary, I always like to have indices on FKs to support flexible querying

matt-bernstein replied:
Yep - I tried adding an index in a concurrent migration, but wasn't able to do it in a way that passed the migration linter

jombooth replied:
It's fine to suppress the migration linter like so: https://github.com/HumanSignal/label-studio/blob/4a1582d55a09a89a93095363eb924a7abb5df94b/label_studio/users/migrations/0010_userproducttour.py#L16 - the index operation will be fast because this is a new column where all the new values will be NULL

jombooth commented on label_studio/data_manager/models.py:
```diff
@@ -86,7 +86,25 @@ class FilterGroup(models.Model):
+        _('index'),
+        null=True,
+        blank=True,
+        default=None,
```
nit - as long as we are making other changes, we can remove this
```suggestion
```

jombooth commented on label_studio/data_manager/models.py:
```diff
@@ -86,7 +86,25 @@ class FilterGroup(models.Model):
+    )
+
+    # `index` is now only meaningful for **root** filters (parent is NULL)
+    index = models.IntegerField(
```
is there a way we could name this field to make it clearer what it really represents? `display_index` or `display_order_index` would both help in this regard, imo

matt-bernstein replied:
This field isn't new, renaming it would have increased change surface area, but yes it's possible. Could do it in a separate chore PR

matt-bernstein committed:
fix child filter persistence

matt-bernstein committed:
wip persist from db

matt-bernstein committed:
wip persist from db

matt-bernstein committed:
wip remove parent_index from BE

matt-bernstein committed:
fix test

matt-bernstein committed:
fix serializers

matt-bernstein committed:
remove parent_index in favor of parent, id

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-45

matt-bernstein commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15830381254)

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15830381254

matt-bernstein committed:
hierarchical filters FE

matt-bernstein committed:
badge styling

matt-bernstein committed:
unstack child filters when applying them to queryset

jombooth committed:
Merge branch 'develop' into fb-ROOT-45

matt-bernstein committed:
consolidate serializing logic

jombooth committed:
Merge branch 'fb-ROOT-45' of github.com:HumanSignal/label-studio into fb-ROOT-45

jombooth committed:
reduce queries in FilterGroupSerializer, reduce duplication in filters application code

jombooth committed:
put the index back

jombooth committed:
test POST and PATCH on the views API - effectively covering view duplication

jombooth committed:
fix broken tests

jombooth committed:
allowlist completed_by

ricardoantoniocm committed:
Refactor Filter components for improved layout and styling

- Updated FilterDropdown to include padding via triggerProps.
- Enhanced FilterInput with new class names for better styling.
- Simplified Filters.scss by removing unused styles and adjusting layout to use flexbox.
- Refined FilterLine component to maintain grid structure and improve child filter rendering.
- Adjusted FilterLine.scss for consistent spacing and alignment.
- Cleaned up String filter component for better readability.

ricardoantoniocm committed:
Merge branch 'fb-ROOT-45' of https://github.com/heartexlabs/label-studio into fb-ROOT-45

ricardoantoniocm committed:
Improves spacing.

ricardoantoniocm committed:
Fixes spacing on sidebar.

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-45

matt-bernstein committed:
revert guards

matt-bernstein committed:
fix guards

matt-bernstein committed:
Revert "fix broken tests"

This reverts commit 22b8f7216497cca0a8329875878ef714c0402b16.

matt-bernstein committed:
Revert "fix test"

This reverts commit f34ed29ac412d8558bb1bb55ce50a58aad7b81f5.

matt-bernstein committed:
wip

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-45

matt-bernstein commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15880045382)

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15880045382

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-45

matt-bernstein commented:
/fmt
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16011901564)

matt-bernstein committed:
merge migrations

matt-bernstein committed:
syntax

matt-bernstein commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16012076391)

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16012076391

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-45

matt-bernstein commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16028122992)

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16028122992

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-45

matt-bernstein committed:
add back views

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-45

matt-bernstein commented:
/fmt
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16503637085)

matt-bernstein committed:
Apply pre-commit linters

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16503637085

matt-bernstein commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16503705384)

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16503705384

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-45

matt-bernstein commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16683483097)

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16683483097

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-45

matt-bernstein committed:
fix partial child filter saving

matt-bernstein committed:
fix style

matt-bernstein commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16686352539)

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16686352539

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-45

matt-bernstein commented on web/libs/datamanager/src/stores/Tabs/tab.js:
```diff
@@ -375,9 +388,37 @@ export const Tab = types
+     * Used internally to materialize join filters.
+     */
+    createChildFilterForType(filterType, parentFilter) {
+      // // HACK: Create a copy of filterType to avoid modifying the protected reference
+      // const filterTypeCopy = { ...filterType };
+      // if (filterType.field.title === "Annotators") {
+      //   filterTypeCopy.field = { ...filterType.field, id: "annotators__completed_by" };
+      // }
+      // filterType = filterTypeCopy;
```
remove

matt-bernstein commented on web/libs/datamanager/src/stores/Tabs/tab_column.jsx:
```diff
@@ -75,6 +75,8 @@ export const TabColumn = types
     target: types.enumeration(["tasks", "annotations"]),
     orderable: types.optional(types.boolean, true),
     help: types.maybeNull(types.string),
+    // List of column aliases whose filters should be joined automatically when a filter is created for this column
+    join_filters: types.maybeNull(types.array(types.string)),
```
rename

matt-bernstein replied:
check SDK docs

matt-bernstein committed:
handle annotator filter in child

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-45

matt-bernstein committed:
join_filters -> child_filter

matt-bernstein committed:
remove debug logs

jombooth committed:
make ChildFilterSerializer into a model serializer so that it Spectacular can understand its fields

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16735511791

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-ROOT-45'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16735511791

jombooth commented on label_studio/data_manager/serializers.py:
```diff
@@ -190,6 +274,8 @@ def update(self, instance, validated_data):
                 if filter_group is None:
                     filter_group = FilterGroup.objects.create(**filter_group_data)
+                    instance.filter_group = filter_group
+                    instance.save()
```
should probably use `update_fields` here

matt-bernstein committed:
fix cutoff connector text

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-45

matt-bernstein committed:
fix remove button

matt-bernstein committed:
revert style change

matt-bernstein committed:
use update_fields

matt-bernstein committed:
fix margin

jombooth commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16762698832)
> Successfully merged:  9 files changed, 10 insertions(+), 18 deletions(-)

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-45

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-45

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16808938594)
> Successfully merged:  3 files changed, 350 insertions(+), 353 deletions(-)

matt-bernstein committed:
Merge branch 'develop' into 'fb-ROOT-45'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16808938594

jombooth reviewed (approved):
LG, awesome!

yyassi-heartex reviewed (approved):
looks good to me!

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16815290383)
> Successfully merged: Already up to date.

matt-bernstein commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16815444605)

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16815403775

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16815464433)
> Successfully merged: Already up to date.

matt-bernstein committed:
lockfile

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio-enterprise/actions/runs/16830032372
